databricks/koalas/base.py

```diff
@@ -41,22 +41,128 @@
 from databricks.koalas.internal import (
     InternalFrame,
+    DEFAULT_SERIES_NAME,
     NATURAL_ORDER_COLUMN_NAME,
     SPARK_DEFAULT_INDEX_NAME,
 )
 from databricks.koalas.spark import functions as SF
 from databricks.koalas.spark.accessors import SparkIndexOpsMethods
 from databricks.koalas.typedef import as_spark_type, spark_type_to_pandas_dtype
-from databricks.koalas.utils import scol_for, validate_axis
+from databricks.koalas.utils import (
+    combine_frames,
+    same_anchor,
+    scol_for,
+    validate_axis,
+    ERROR_MESSAGE_CANNOT_COMBINE,
+)
 from databricks.koalas.frame import DataFrame
 
 if TYPE_CHECKING:
     from databricks.koalas.indexes import Index
     from databricks.koalas.series import Series
 
 
+def need_alignment_for_column_op(self: "IndexOpsMixin", other: "IndexOpsMixin") -> bool:
+    from databricks.koalas.series import Series
+
+    if isinstance(self, Series) and isinstance(other, Series):
+        return not same_anchor(self, other)
+    else:
+        return self._internal.spark_frame is not other._internal.spark_frame
+
+
+def align_diff_series(func, this_series: "Series", *args) -> "Series":
+    from databricks.koalas.series import Series
+
+    cols = [arg for arg in args if isinstance(arg, Series)]
+    combined = combine_frames(this_series.to_frame(), *cols, how="full")
+
+    return column_op(func)(
+        combined["this"]._kser_for(combined["this"]._internal.column_labels[0]),
+        *[combined["that"]._kser_for(label) for label in combined["that"]._internal.column_labels]
+    )
+
+
+def align_diff_index(func, this_index: "Index", *args) -> "Index":
+    from databricks.koalas.indexes import Index
+    from databricks.koalas.series import Series
+
+    with option_context("compute.default_index_type", "distributed-sequence"):
+        # Directly using Series from both self and other seems causing
+        # some exceptions when 'compute.ops_on_diff_frames' is enabled.
+        # Working around for now via using frame.
+        return (
+            cast(
+                Series,
+                column_op(func)(
+                    this_index.to_series().reset_index(drop=True),
+                    *[
+                        arg.to_series().reset_index(drop=True) if isinstance(arg, Index) else arg
+                        for arg in args
+                    ]
+                ),
+            )
+            .sort_index()
+            .to_frame(DEFAULT_SERIES_NAME)
+            .set_index(DEFAULT_SERIES_NAME)
+            .index.rename(this_index.name)
+        )
+
+
+def align_diff_index_ops(func, this_index_ops: "IndexOpsMixin", *args) -> "Series":
+    from databricks.koalas.indexes import Index
+    from databricks.koalas.series import Series, first_series
+
+    cols = [arg for arg in args if isinstance(arg, IndexOpsMixin)]
+
+    with option_context("compute.default_index_type", "distributed-sequence"):
+        if isinstance(this_index_ops, Series):
```

Comment:
@ueshin, can we briefly explain what `align_diff_series`, `align_diff_index` and `align_diff_index_ops` do in terms of input/output? From reading the names, looks like we can just use `align_diff_index_ops` instead of `align_diff_series` and `align_diff_index`.